File: Infectious Disease Case Detection and Modeling Framework/scripts/python/bivariate_analysis.py
"""
Bivariate Analysis Module
"""
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import sys
import os
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from data_loader import load_data, clean_data

def bivariate_analysis(df_clean, var1='Incidence_Rate', var2='Annual_Count'):
    """
    Perform bivariate analysis between two variables
    
    Parameters:
    -----------
    df_clean : DataFrame
        Cleaned dataset
    var1 : str
        First variable
    var2 : str
        Second variable
    """
    print(f"=== BIVARIATE ANALYSIS: {var1} vs {var2} ===\n")
    
    # Remove missing values
    data = df_clean[[var1, var2]].dropna()
    
    # Correlation analysis
    correlation = data[var1].corr(data[var2])
    print(f"1. Correlation Coefficient: {correlation:.4f}")
    
    # Pearson correlation test
    pearson_r, pearson_p = stats.pearsonr(data[var1], data[var2])
    print(f"   Pearson correlation: r={pearson_r:.4f}, p-value={pearson_p:.4f}")
    
    # Spearman correlation
    spearman_r, spearman_p = stats.spearmanr(data[var1], data[var2])
    print(f"   Spearman correlation: r={spearman_r:.4f}, p-value={spearman_p:.4f}")
    
    # Visualizations
    fig, axes = plt.subplots(2, 2, figsize=(15, 12))
    
    # Scatter plot
    axes[0, 0].scatter(data[var1], data[var2], alpha=0.5)
    axes[0, 0].set_xlabel(var1)
    axes[0, 0].set_ylabel(var2)
    axes[0, 0].set_title(f'Scatter Plot: {var1} vs {var2}')
    
    # Add regression line
    z = np.polyfit(data[var1], data[var2], 1)
    p = np.poly1d(z)
    axes[0, 0].plot(data[var1], p(data[var1]), "r--", alpha=0.8, label=f'Trend line')
    axes[0, 0].legend()
    
    # Hexbin plot
    axes[0, 1].hexbin(data[var1], data[var2], gridsize=20, cmap='Blues')
    axes[0, 1].set_xlabel(var1)
    axes[0, 1].set_ylabel(var2)
    axes[0, 1].set_title(f'Hexbin Plot: {var1} vs {var2}')
    
    # 2D histogram
    axes[1, 0].hist2d(data[var1], data[var2], bins=30, cmap='Blues')
    axes[1, 0].set_xlabel(var1)
    axes[1, 0].set_ylabel(var2)
    axes[1, 0].set_title(f'2D Histogram: {var1} vs {var2}')
    
    # Joint plot (simplified)
    sns.scatterplot(data=data, x=var1, y=var2, ax=axes[1, 1])
    axes[1, 1].set_title(f'Scatter with Density: {var1} vs {var2}')
    
    plt.tight_layout()
    plt.savefig(f'bivariate_{var1}_vs_{var2}.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    return {
        'correlation': correlation,
        'pearson_r': pearson_r,
        'pearson_p': pearson_p,
        'spearman_r': spearman_r,
        'spearman_p': spearman_p
    }

def categorical_bivariate_analysis(df_clean, cat_var='Trend', num_var='Incidence_Rate'):
    """
    Analyze relationship between categorical and numerical variables
    
    Parameters:
    -----------
    df_clean : DataFrame
        Cleaned dataset
    cat_var : str
        Categorical variable
    num_var : str
        Numerical variable
    """
    print(f"=== BIVARIATE ANALYSIS: {cat_var} vs {num_var} ===\n")
    
    if cat_var not in df_clean.columns:
        print(f"Variable {cat_var} not found in dataset")
        return
    
    data = df_clean[[cat_var, num_var]].dropna()
    
    # Group statistics
    group_stats = data.groupby(cat_var)[num_var].agg(['count', 'mean', 'median', 'std'])
    print("1. Group Statistics:")
    print(group_stats)
    
    # Visualizations
    fig, axes = plt.subplots(2, 2, figsize=(15, 12))
    
    # Box plot
    categories = data[cat_var].unique()
    box_data = [data[data[cat_var] == cat][num_var].values for cat in categories if pd.notna(cat)]
    axes[0, 0].boxplot(box_data, labels=[cat for cat in categories if pd.notna(cat)])
    axes[0, 0].set_xlabel(cat_var)
    axes[0, 0].set_ylabel(num_var)
    axes[0, 0].set_title(f'Box Plot: {num_var} by {cat_var}')
    axes[0, 0].tick_params(axis='x', rotation=45)
    
    # Violin plot
    sns.violinplot(data=data, x=cat_var, y=num_var, ax=axes[0, 1])
    axes[0, 1].set_title(f'Violin Plot: {num_var} by {cat_var}')
    axes[0, 1].tick_params(axis='x', rotation=45)
    
    # Bar plot of means
    means = data.groupby(cat_var)[num_var].mean()
    axes[1, 0].bar(means.index, means.values)
    axes[1, 0].set_xlabel(cat_var)
    axes[1, 0].set_ylabel(f'Mean {num_var}')
    axes[1, 0].set_title(f'Mean {num_var} by {cat_var}')
    axes[1, 0].tick_params(axis='x', rotation=45)
    
    # Strip plot
    sns.stripplot(data=data, x=cat_var, y=num_var, ax=axes[1, 1], alpha=0.5)
    axes[1, 1].set_title(f'Strip Plot: {num_var} by {cat_var}')
    axes[1, 1].tick_params(axis='x', rotation=45)
    
    plt.tight_layout()
    plt.savefig(f'bivariate_{cat_var}_vs_{num_var}.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # ANOVA test
    groups = [data[data[cat_var] == cat][num_var].values 
              for cat in categories if pd.notna(cat) and len(data[data[cat_var] == cat]) > 0]
    if len(groups) > 2:
        f_stat, p_value = stats.f_oneway(*groups)
        print(f"\n2. ANOVA Test:")
        print(f"   F-statistic: {f_stat:.4f}")
        print(f"   p-value: {p_value:.4f}")
        print(f"   Result: {'Significant difference' if p_value < 0.05 else 'No significant difference'}")
    
    return group_stats

if __name__ == "__main__":
    df = load_data()
    df_clean = clean_data(df)
    
    # Numerical vs Numerical
    result1 = bivariate_analysis(df_clean, 'Incidence_Rate', 'Annual_Count')
    
    # Categorical vs Numerical
    if 'Trend' in df_clean.columns:
        result2 = categorical_bivariate_analysis(df_clean, 'Trend', 'Incidence_Rate')

File: Infectious Disease Case Detection and Modeling Framework/scripts/r/bivariate_analysis.R
# Bivariate Analysis Module
# R script for bivariate statistical analysis

library(ggplot2)
library(dplyr)
library(corrplot)
library(gridExtra)
library(rlang)
source("data_loader.R")

# Function for bivariate analysis (numerical vs numerical)
# Perform bivariate analysis between two variables
# Parameters: df_clean - cleaned dataset, var1/var2 - variables to analyze
bivariate_analysis <- function(df_clean, var1 = "Incidence_Rate", var2 = "Annual_Count") {
  cat("=== BIVARIATE ANALYSIS:", var1, "vs", var2, "===\n\n")
  
  # Remove missing values
  data <- df_clean %>%
    select(all_of(c(var1, var2))) %>%
    filter(!is.na(.data[[var1]]) & !is.na(.data[[var2]]))
  
  # Correlation analysis
  correlation <- cor(data[[var1]], data[[var2]])
  cat("1. Correlation Coefficient:", correlation, "\n")
  
  # Pearson correlation test
  pearson_test <- cor.test(data[[var1]], data[[var2]], method = "pearson")
  cat("   Pearson correlation: r =", pearson_test$estimate, 
      ", p-value =", pearson_test$p.value, "\n")
  
  # Spearman correlation
  spearman_test <- cor.test(data[[var1]], data[[var2]], method = "spearman")
  cat("   Spearman correlation: rho =", spearman_test$estimate, 
      ", p-value =", spearman_test$p.value, "\n")
  
  # Create visualizations
  p1 <- ggplot(data, aes_string(x = var1, y = var2)) +
    geom_point(alpha = 0.5) +
    geom_smooth(method = "lm", se = TRUE, color = "red") +
    labs(title = paste("Scatter Plot:", var1, "vs", var2),
         x = var1, y = var2) +
    theme_minimal()
  
  p2 <- ggplot(data, aes_string(x = var1, y = var2)) +
    geom_hex(bins = 30) +
    scale_fill_gradient(low = "white", high = "steelblue") +
    labs(title = paste("Hexbin Plot:", var1, "vs", var2),
         x = var1, y = var2) +
    theme_minimal()
  
  # Combine plots
  grid.arrange(p1, p2, ncol = 2)
  
  return(list(
    correlation = correlation,
    pearson_r = pearson_test$estimate,
    pearson_p = pearson_test$p.value,
    spearman_r = spearman_test$estimate,
    spearman_p = spearman_test$p.value
  ))
}

# Function for categorical vs numerical analysis
# Analyze relationship between categorical and numerical variables
# Parameters: df_clean - cleaned dataset, cat_var - categorical variable, num_var - numerical variable
categorical_bivariate_analysis <- function(df_clean, cat_var = "Trend", num_var = "Incidence_Rate") {
  cat("=== BIVARIATE ANALYSIS:", cat_var, "vs", num_var, "===\n\n")
  
  if (!cat_var %in% colnames(df_clean)) {
    cat("Variable", cat_var, "not found in dataset\n")
    return(NULL)
  }
  
  data <- df_clean %>%
    select(all_of(c(cat_var, num_var))) %>%
    filter(!is.na(.data[[cat_var]]) & !is.na(.data[[num_var]]))
  
  # Group statistics
  group_stats <- data %>%
    group_by(.data[[cat_var]]) %>%
    summarise(
      count = n(),
      mean = mean(.data[[num_var]], na.rm = TRUE),
      median = median(.data[[num_var]], na.rm = TRUE),
      sd = sd(.data[[num_var]], na.rm = TRUE),
      .groups = "drop"
    )
  
  cat("1. Group Statistics:\n")
  print(group_stats)
  
  # Create visualizations
  p1 <- ggplot(data, aes_string(x = cat_var, y = num_var)) +
    geom_boxplot(fill = "steelblue", alpha = 0.7) +
    labs(title = paste("Box Plot:", num_var, "by", cat_var),
         x = cat_var, y = num_var) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  p2 <- ggplot(data, aes_string(x = cat_var, y = num_var)) +
    geom_violin(fill = "steelblue", alpha = 0.7) +
    labs(title = paste("Violin Plot:", num_var, "by", cat_var),
         x = cat_var, y = num_var) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  p3 <- ggplot(group_stats, aes_string(x = cat_var, y = "mean")) +
    geom_bar(stat = "identity", fill = "steelblue", alpha = 0.7) +
    labs(title = paste("Mean", num_var, "by", cat_var),
         x = cat_var, y = paste("Mean", num_var)) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  p4 <- ggplot(data, aes_string(x = cat_var, y = num_var)) +
    geom_jitter(alpha = 0.5, width = 0.2) +
    labs(title = paste("Strip Plot:", num_var, "by", cat_var),
         x = cat_var, y = num_var) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  # Combine plots
  grid.arrange(p1, p2, p3, p4, ncol = 2, nrow = 2)
  
  # ANOVA test
  if (length(unique(data[[cat_var]])) > 2) {
    anova_result <- aov(data[[num_var]] ~ data[[cat_var]])
    anova_summary <- summary(anova_result)
    
    cat("\n2. ANOVA Test:\n")
    print(anova_summary)
  }
  
  return(group_stats)
}

# Main execution
if (!interactive()) {
  df <- load_data()
  df_clean <- clean_data(df)
  
  # Numerical vs Numerical
  result1 <- bivariate_analysis(df_clean, "Incidence_Rate", "Annual_Count")
  
  # Categorical vs Numerical
  if ("Trend" %in% colnames(df_clean)) {
    result2 <- categorical_bivariate_analysis(df_clean, "Trend", "Incidence_Rate")
  }
}

